modules/bullet/collision_object_bullet.h

```diff
@@ -44,13 +44,35 @@ class btCollisionShape;
 
 class CollisionObjectBullet : public ShapeOwnerBullet {
 protected:
-	class HackedBasis : public Basis {
+	class EnhancedBasis : public Basis {
 	public:
 		void get_ABS_scale(Vector3 &out_vec) const {
 			out_vec[0] = Vector3(elements[0][0], elements[1][0], elements[2][0]).length();
 			out_vec[1] = Vector3(elements[0][1], elements[1][1], elements[2][1]).length();
 			out_vec[2] = Vector3(elements[0][2], elements[1][2], elements[2][2]).length();
 		}
+
+		real_t get_uniform_scale() const {
```

Comment:
Not all non-uniform scale leads to shearing actually. Bullet should be handle a non-uniform scale in such a situation for example: Let's say parent and grandparent have uniform scale, and only the child has non-uniform scale. The global transform can then be written as: Tg = R3.S3.R2.S2.R1.S1 = (R3.R2.R1).(S3.S2.S1). The R3.R2.R1 part would be the global rotation and S3.S2.S1 would be local scaling.

If you use `Basis::rotref_posscale_decomposition` in PR #10641 (I believe it will be merged soon), it actually checks for shearing before attempting at a decomposition, and prints out a warning in the CLI when shearing is present.

Reply:
#10641 is actually merged just now, so you should be able to use it directly

Reply:
To support non-uniform scaling I must do a lot more work with transformation:
- I've to check if the scaling induce shearing in the local shape. In this case there is another problem since I need to alert the editor that this transform induce shearing.
- I have to rotate the global scaling to be applied to the local shape, and this also can bring some float point precision error. 
- Currently in Bullet only the box shape support at 100% non-uniform scaling.

Thus I prefer to not support it since the improvements that it bring are less then the problems.
This "restriction" make the code and life easier, if the developer / artist need a particular shape dimentsion, he can just insert the corrects dimensions directly to the shape instead of control it using the scale (that at the end is the right thing to do).

Reply:
> I've to check if the scaling induce shearing in the local shape. In this case there is another problem since I need to alert the editor that this transform induce shearing.

As you can see in #10641, this doesn't require an additional check.

> I have to rotate the global scaling to be applied to the local shape, and this also can bring some float point precision error.

You don't have to rotate anything, again, as you can see in #10641.

By the way, such floating precision errors are always there. Every physics tick induces floating point errors *in addition* to the finite difference errors.

> Currently in Bullet only the box shape support at 100% non-uniform scaling.

You can still do that with #10641.

What I'm trying to convince you is, such changes (whatever you need, although I'm pretty sure #10641 is enough to do everything you want to do) should be done in Basis, rather than one a workaround inherited class.

And since #10641 is merged, you can use it.

Reply:
>>I've to check if the scaling induce shearing in the local shape. In this case there is another problem since I need to alert the editor that this transform induce shearing.

> As you can see in #10641, this doesn't require an additional check.

The check is required to alert the developer / artist that this transform is bad, If this check is not done in my side must be done in godot side (and I don't think it's a good thing).

>>I have to rotate the global scaling to be applied to the local shape, and this also can bring some float point precision error.

>You don't have to rotate anything, again, as you can see in #10641.

The problem here is that you didn't understand well how it works, I need to apply the shape scale in local space. When I receive a scale of x1, y2, z1 (that is not-uniform scale and is in a global space) it take different meaning in base of the local shape rotation.
For example
- if the shape locally is rotated by 45° the shape will be stretched in the rendering engine, and thereis no way to stretch it in godot.
- if the shape is rotated by 90° I need to scale the X axis of shape 2 times and not the Y axis.

As you can see with not-uniform scale I've to do more work to check if it fit or not, and this work increase the floating point error. It's not only a problem of the orthonormalize that I'm forced to use it to be sure the transform perfect for Bullet.

> What I'm trying to convince you is, such changes (whatever you need, although I'm pretty sure #10641 is enough to do everything you want to do) should be done in Basis, rather than in a workaround inherited class.

You can't think my class is an Hack just because is not in the main class, I did it to avoid a PR and add a feature that will not be used by anyone since it's required only for the physics engine.

However please download this PR and apply your code, you'll see that this doesn't work for the reason written above.

However I want to point out that your feedback is very appreciated, and for this I'm saying you a big thanks!!

Reply:
#10641 handles all of it, and I understand what you're doing, don't worry.

The reason it "doesn't work" is because #10641 returns a zero scale whenever it gets a bad input rather than trying to pick some random scale.

Also you don't need to use a slow operation like orthonormalize there at all

Anyway feel free to submit your code with it, I'll submit my own PR after it gets merged